addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,69 +1,128 @@
-        end_date = from_date + relativedelta(day=calendar.monthrange(from_date.year, from_date.month)[1])
+    def _consolidate_cron(self):
+        """Consolidate the trackings 2 months ago. Used by a cron to cleanup tracking records."""
+        from_date = date_utils.start_of(fields.Datetime.today(), 'month') - relativedelta(months=2)
+        return self._process_consolidate(from_date)
+
+    def _process_consolidate(self, from_date, end_date=None):
+        """Consolidate the karma trackings, from the given date month.
+
+        The consolidation keeps, for each user, the oldest "old_value" and the most recent
+        "new_value", creates a new karma tracking with those values and removes all karma
+        trackings between those dates.
+        """
+        self.env['gamification.karma.tracking'].flush_model()
+
+        from_date = date_utils.start_of(from_date, 'month')
+        if not end_date:
+            end_date = date_utils.end_of(date_utils.end_of(from_date, 'month'), 'day')
+
         select_query = """
-SELECT user_id,
-(
-    SELECT old_value from gamification_karma_tracking old_tracking
-    WHERE old_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date ASC LIMIT 1
-), (
-    SELECT new_value from gamification_karma_tracking new_tracking
-    WHERE new_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date DESC LIMIT 1
-)
-FROM gamification_karma_tracking
-WHERE tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-AND consolidated IS NOT TRUE
-GROUP BY user_id """
+        WITH old_tracking AS (
+            SELECT DISTINCT ON (user_id) user_id, old_value, tracking_date
+              FROM gamification_karma_tracking
+             WHERE tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND consolidated IS NOT TRUE
+          ORDER BY user_id, tracking_date ASC, id ASC
+        )
+            INSERT INTO gamification_karma_tracking (
+                            user_id,
+                            old_value,
+                            new_value,
+                            tracking_date,
+                            origin_ref,
+                            consolidated,
+                            reason)
+            SELECT DISTINCT ON (nt.user_id)
+                            nt.user_id,
+                            ot.old_value AS old_value,
+                            nt.new_value AS new_value,
+                            ot.tracking_date AS from_tracking_date,
+                            %(origin_ref)s AS origin_ref,
+                            TRUE,
+                            format(%(reason)s, ot.tracking_date::DATE, nt.tracking_date::DATE)
```

Comment:
Should we just say "consolidate from 2023-03-01 to 2023-03-31" instead of taking the dates from the tracking?

I feel like it's much clearer, we don't need to print the dates of the actual oldest and newest trackings, and it also makes it more simple as you don't need to format anything here.